contracts/l2/FuseController.sol

```diff
@@ -0,0 +1,309 @@
+import "./IFuseController.sol";
+import "./IControllerUpgrade.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+error Unauthorised(bytes32 node, address addr);
+error CannotUpgrade();
+error nameExpired(bytes32 node);
+
+/**
+ * @dev A simple ENS registry controller. Names are permanently owned by a single account.
+ *      Name data is structured as follows:
+ *       - Byte 0: controller (address)
+ *       - Byte 20: owner (address)
+ *       - Byte 40: resolver (address)
+ *       _ Byte 60: expiry (uint64)
+ *       - Byte 68: fuses (uint96)
+ *       - Byte 80: renewalController (address)
+ */
+contract FuseController is Ownable, IFuseController {
+    L2Registry immutable registry;
+
+    IControllerUpgrade upgradeContract;
+
+    // A struct to hold the unpacked data
+    struct TokenData {
+        address owner;
+        address resolver;
+        uint64 expiry;
+        uint96 fuses;
+        address renewalController;
+    }
+
+    constructor(L2Registry _registry) {
+        registry = _registry;
+    }
+
+    /*************************
+     * IController functions *
+     *************************/
+
+    function ownerOfWithData(
+        bytes calldata tokenData
+    ) external pure returns (address) {
+        (address owner, , , , ) = _unpack(tokenData);
+        return owner;
+    }
+
+    function ownerOf(bytes32 node) external view returns (address) {
+        //get the tokenData
+        bytes memory tokenData = registry.getData(uint256(node));
+        (address owner, , , , ) = _unpack(tokenData);
+        return owner;
+    }
+
+    function safeTransferFrom(
+        bytes calldata tokenData,
+        address operator,
+        address from,
+        address to,
+        uint256 /*id*/,
+        uint256 value,
+        bytes calldata /*data*/,
+        bool operatorApproved
+    ) external view returns (bytes memory) {
+        TokenData memory td;
+
+        (
+            td.owner,
+            td.resolver,
+            td.expiry,
+            td.fuses,
+            td.renewalController
+        ) = _unpack(tokenData);
+
+        require(value == 1);
+        require(from == td.owner);
+        require(operator == td.owner || operatorApproved);
+
+        return
+            _pack(
+                address(this),
+                to,
+                td.resolver,
+                td.expiry,
+                td.fuses,
+                td.renewalController
+            );
+    }
+
+    function balanceOf(
+        bytes calldata tokenData,
+        address _owner,
+        uint256 /*id*/
+    ) external pure returns (uint256) {
+        (address owner, , , , ) = _unpack(tokenData);
+        return _owner == owner ? 1 : 0;
+    }
+
+    function resolverFor(
+        bytes calldata tokenData
+    ) external pure returns (address) {
+        (, address resolver, , , ) = _unpack(tokenData);
+        return resolver;
+    }
+
+    function expiryOf(bytes32 node) external view returns (uint64) {
```

Comment:
A generic `data` method may be more useful here, as well as being more gas efficient.